### PR TITLE
previous PR fixes, updated vulcan metrics from original vulcan, and more

### PR DESCRIPTION
**Pull request by priyatharsan**
https://github.com/Aifred-Health/Vulcan2/pull/1 fixes:
- [x] Break down `config['units']` to two: one for `_conv_` and the other for `_dense_`
- [x]  add more detail within the units to specify filter count, filter size, stride, dilation factor, etc
- [x] Generalize `get output size`
- [x] Move back `build_conv_network` and `build_dense_network` to respective network classes
- [x] Move the notebooks to `examples`
- [x] `build_dense_network` must be instantiated in ConvNet only when needed

https://github.com/Aifred-Health/Vulcan2/pull/3 fixes:
- [x] We are missing a few metrics here like sensitivity, specificity, npc, ppv, etc that can be found in the original Vulcan
- [x] use sklearn to generate the confusion matrix (renaming the sklearn import)
- [x] sklearn roc and abc calculation work better than scipy (renaming the sklearn import)

### CODE REVIEW

**Comment by priyatharsan** [comment]
- [ ] One of the next milestone would be to integrate the functionalities of TabularDataset, Logger and Models/Networks
- [ ] Add visualization class
- [ ] Clean up Layers class and fix the Weight and Bias initializers

@rfratila  and @nirtiac please add next milestones in this comment (please edit this) besides the new PR-fix requests/changes/suggestions.

File: vulcanai2/models/metrics.py [hunk 1]
```diff
@@ -1,66 +1,371 @@
...
-        hist = np.bincount(
-                            num_class * target[mask].astype(int) + prediction[mask],
-                            minlength=num_class ** 2,
-                            ).reshape(num_class, num_class)
-        return hist
+        self.list_classes = list(range(num_class))
+        self.use_unlabeled = use_unlabeled
+        self.mat_start_idx = 1 if not self.use_unlabeled else 0
+
+    def reset(self):
```
Comment (rfratila) [inline vulcanai2/models/metrics.py]: Why do we reset the metrics matrix's? why not just overwrite them?

File: vulcanai2/models/metrics.py [hunk 2]
```diff
@@ -1,66 +1,371 @@
...
+        self.mIoU = 0
+    
     def update(self, predictions, targets):
-        for prediction, target in zip(predictions, targets):
-            self.confusion_matrix += self._fast_hist(
-                prediction.flatten(), target.flatten(), self.num_class)
+        if not(isinstance(predictions, np.ndarray)) or not(isinstance(targets, np.ndarray)):
+            print("Expected ndarray")
+
+        elif len(targets.shape) == 3:        # batched spatial target
```
Comment (rfratila) [inline vulcanai2/models/metrics.py]: is there a way that we can reduce this else if ladder?

File: examples/test.py [hunk 3]
```diff
@@ -39,20 +40,51 @@
...
-network_conv_config = {
-    'mode': 'conv',
-    'units': ([ [1, 16, (5, 5), 2, 0],
-                [16, 32, (5, 5), 1, 0],
-                [32, 64, (5, 5), 1, 0] ],
-              [ 100 ]),
-    'dropouts': [0.2],
+conv_net_config = {
+    'conv_units': [
+                    dict(
```
[the remark this answers is not in the document]
Comment (priyatharsan) [reply examples/test.py]: 👍

File: vulcanai2/models/basenetwork.py [hunk 4]
```diff
@@ -110,15 +113,23 @@ def stopping_rule(self, value):
...
 
-    def get_conv_output_size(self):
+    def get_flattened_size(self, network):
         """
-        Helper function to calculate the size of the flattened 
-        features after the last conv layer
+        Returns the flattened output size of the conv network's last layer
         """
         with torch.no_grad():
             x = torch.ones(1, *self.in_dim)
```
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: Yeah we could either use something like:
```
x = Variable(torch.rand(1, *self.in_dim))
x  = network(x)
n_size = x.data.view(1, -1).size(1)
return n_size
```
or:
```
x = torch.ones(1, *self.in_dim)
x = network(x)
return x.numel() # which outputs num of elements
```
either way we will have to do a minor forward pass to the network to get the size: https://discuss.pytorch.org/t/inferring-shape-via-flatten-operator/138/4.

[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: I felt the latter approach was simpler

File: vulcanai2/models/basenetwork.py [hunk 5]
```diff
@@ -188,99 +199,112 @@ def _init_trainer(self):
...
-        self.print_iu = print_iu
 
         self._init_trainer()
 
         for epoch in trange(self.epoch, epochs, desc='Epoch: ', ncols=80):
-            self.epoch = epoch
-            self.train_epoch()                    
+            train_loss, train_acc= self.train_epoch()
+            valid_loss, acc, avg_acc, iou, miou, conf_mat = self.validate()
+            tqdm.write("\n Epoch {}:\nTrain Loss: {:.6f} | Test Loss: {:.6f} | Train Acc: {:.2f}% | Test Acc: {:.2f}%".format(epoch, train_loss, valid_loss, train_acc*100, avg_acc*100))
```
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: no problem! may we could have some sort of option that can be added to the logger.

File: vulcanai2/models/basenetwork.py [hunk 6]
```diff
@@ -188,99 +199,112 @@ def _init_trainer(self):
...
+                    pbar.update(len(self.train_loader.dataset) - int(batch_idx*len(data)))
 
+            _, acc = self.metrics.get_accuracy(predictions, targets)
+
+        pbar.close()
+        return (train_loss*len(data)/len(self.train_loader.dataset), acc)
+            
     def validate(self):
-        training = self.training
         self.eval()  # Set model to evaluate mode
```
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: just adding for ref: https://stackoverflow.com/questions/48146926/whats-the-meaning-of-function-eval-in-torch-nn-module

File: vulcanai2/models/metrics.py [hunk 7]
```diff
@@ -1,66 +1,371 @@
...
 import logging
 logger = logging.getLogger(__name__)
 
 class Metrics(object):
 
-    def __init__(self, num_class):
+    def __init__(self, num_class, use_unlabeled=False):
+        self.mat = np.zeros((num_class, num_class), dtype=np.float)
+        self.valids = np.zeros((num_class), dtype=np.float)
+        self.IoU = np.zeros((num_class), dtype=np.float)
```
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: Intersectio over Unions. Inspired from : https://www.pyimagesearch.com/2016/11/07/intersection-over-union-iou-for-object-detection/

File: vulcanai2/models/metrics.py (hunk 1 above)
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: thought it would be modular having reset as an option and helpful while logging. Not a good idea?

File: vulcanai2/models/basenetwork.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: sure as long as it gets the job done it should be fine for now. Will this generalize well to multiple inputs for concatenating into a deeper layer?

File: vulcanai2/models/basenetwork.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: well it's more of a consistency thing than anything else. It doesn't add more info to the user by converting it but all of our other metrics are in the range(0,1) so this would just make it a more cohesive set of metrics.

File: vulcanai2/models/basenetwork.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: sounds good. let's include that link in-line

File: vulcanai2/models/metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/metrics.py]: thanks for linking. however, for our problem space it's not as important as the other metrics we are calculating. once we put a larger emphasis on images we can include this but it's not a very important metric for now.

File: vulcanai2/models/metrics.py (hunk 1 above)
Comment (rfratila) [reply vulcanai2/models/metrics.py]: I mean... if you can think of a key case that requires resetting the metric matrices but I can't think of one. it may be redundant when the logger could overwrite with the most recent test. idk what did you plan on using it for?

File: vulcanai2/models/basenetwork.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: I think it shouldn't be a issue, the main idea is to get the size from the last layer alone (super-effective when automating the conv-dense combination) For example, `(65,4,4)` conv to `1024` dense. May be once we start testing it out with an actual problem of multiple input concatenation.

File: vulcanai2/models/basenetwork.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/basenetwork.py]: I just realized we are sticking with a (0,1) range. I will change that.

File: vulcanai2/models/basenetwork.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: sure. let's deal with it if it comes up while setting up the unit tests

File: vulcanai2/models/metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: Cool I can remove! I was just adding metrics on a broader perspective.

[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/metrics.py]: we might be able to automate the metrics later depending on what the model layers consist of. we can keep it for now. the more the merrier when it comes to metrics :)
